### PR TITLE
ML-aware Irises demo.

### PR DESCRIPTION
**Pull request by dkorolev**
Yo!

After https://github.com/C5T/Current/pull/708.

Check out this visualization from Wikipedia: https://upload.wikimedia.org/wikipedia/commons/5/56/Iris_dataset_scatterplot.svg

Here's the output of `step2`, `http://localhost:3000`:

image

And here's the output of `step3`, `http://localhost:3001`:

image

All 24 plots are clickable.

Would appreciate it if you could test this code on Mac.

Thanks!
Dima

### CODE REVIEW

**Comment by dkorolev** [comment]
And here's what `step3` prints:

```
dima@dima-x1:~/github/dkorolev/Current/examples/Iris@iris_ml $ ./.current/step3_optimize 
Read 150 flowers.
Prior to the optimization, the objective function is: 0.333333
After training the descriptive model, the objective function is: 0.843853
The intermediate parameters vector is: [5.005999988845246,3.4279999923614667,1.4619999967422586,0.2459999994518438,1.8871036486920583,5.936000000620659,2.7700000002896267,4.260000000445418,1.3260000001386446,1.1836343729634509,6.588000007965216,2.974000003595713,5.552000006712642,2.026000002449534,0.8317198309803668]
After training the descriminant model, the objective function is: 0.961264
The optimal parameters vector is: [4.999462187493568,3.4925978168042359,1.3549371276891605,0.18655899315424244,2.0196086101019876,6.41786824924614,3.3447018159313979,4.439485001872361,0.47226107045903339,1.3943706713622182,6.140214283187905,2.4975084761787857,5.683708049537103,2.882915350863361,1.2728438406036389]
Starting the server on http://localhost:3001
```

The `0.84` and `0.94` are the respective accuracies, normalized back from log-losses.

Context: What this `examples/Iris` becomes now is a relatively full-scale demo of how can Current::FnCAS be used to solve an end-to-end optimization & visualization problem. I plan to retire `test.cc` from that directory, and replace it by `step4` introducing `Storage`, and `step5` making the learning real-time.

Let's sync up my morning your evening.
Dima

**Comment by dkorolev** [comment]
And here's the output of the descriptive model only:

```
./.current/step3_optimize  --descriptive_only
Read 150 flowers.
Prior to the optimization, the objective function is: 0.333333
After training the descriptive model, the objective function is: 0.843853
The intermediate parameters vector is: [5.005999988845246,3.4279999923614667,1.4619999967422586,0.2459999994518438,1.8871036486920583,5.936000000620659,2.7700000002896267,4.260000000445418,1.3260000001386446,1.1836343729634509,6.588000007965216,2.974000003595713,5.552000006712642,2.026000002449534,0.8317198309803668]
Starting the server on http://localhost:3001
```

image

**Comment by sompylasar** [comment]
Oh finally the green and blue are in place.

**Comment by dkorolev** [comment]
@mzhurovich -- and here are the training plots generated by the past three commits.

training-plots

File: FnCAS/fncas/optimize.h [hunk 1]
```diff
@@ -309,6 +374,7 @@ template <OptimizationDirection DIRECTION>
 struct OptimizeImpl<GradientDescentOptimizerSelector, DIRECTION> {
   template <typename ORIGINAL_F, typename F, typename G>
   static OptimizationResult RunOptimize(const Optimizer<ORIGINAL_F, DIRECTION>& super,
+                                        const ORIGINAL_F& original_f,
```
Comment (mzhurovich) [inline FnCAS/fncas/optimize.h]: Why do you need `original_f` here? Just to use in `TrackIteration`? What's the difference between "original" and "non-original" f? Do they result in different values for the same point?

**Comment by dkorolev** [comment]
Yes, the original can also compute and track other values, while the
blueprint or compiled one has that extras erased.

File: examples/Iris/step1_analyze_tsv.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+#
+# Step 1: Analyze the TSV file, generate its description, JSON representation, and the schema header.
+
+(cd data &&
+ rm -f rm -f dataset.description dataset.json dataset.h &&
```
Comment (mzhurovich) [inline examples/Iris/step1_analyze_tsv.sh]: Too many `rm -f`'s :-/

File: examples/Iris/step3_optimize.cc [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+template <typename T>
+struct WeightsComputer {
+  T inv_r_squared[3];
+  T k[3];
+
+  // For a 4D Gaussian, `exp(-(distance_squared / radius_squared))`, the normalization coefficient
+  // is `radius_squared ^ (-2)`. As an implementation detail, use inverted values.
+  explicit WeightsComputer(const std::vector<T>& x) {
+    for (size_t c = 0; c < 3; ++c) {
+      inv_r_squared[c] = fncas::exp(x[c * 5 + 4]);
```
Comment (mzhurovich) [inline examples/Iris/step3_optimize.cc]: If you store radius as `-log(r)`, this one is just `inv_r`, where is the `squared` part?

File: examples/Iris/step1_analyze_tsv.sh (hunk 2 above)
Comment (dkorolev) [inline examples/Iris/step1_analyze_tsv.sh]: :+1: 444cf51

File: examples/Iris/step3_optimize.cc (hunk 3 above)
Comment (dkorolev) [inline examples/Iris/step3_optimize.cc]: Right...
Right!
Right.

The optimization code has been correct though, despite the wrong comment. I've elaborated on the math quite a bit; and fixed the visualization part.

PTAL!

Thanks,
Dima

File: examples/Iris/step3_optimize.cc [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+
+    ModelVisualizer(const std::vector<double>& parameters) : parameters(parameters) {}
+
+    void AddPlots(current::gnuplot::GNUPlot& plot, size_t dimension_x, size_t dimension_y) const {
+      using namespace current::gnuplot;
+
+      for (size_t c = 0; c < 3; ++c) {
+        const double x = parameters[c * 5 + dimension_x];
+        const double y = parameters[c * 5 + dimension_y];
+        const double r = std::exp(parameters[c * 5 + 4] / 2) * 0.25;  // Just the visualization coeffcient. -- D.K.
```
Comment (mzhurovich) [inline examples/Iris/step3_optimize.cc]: s/coeffcient/coefficient/

File: examples/Iris/step3_optimize.cc [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+  } catch (const std::out_of_range&) {
+    return Response("Invalid dimension.", HTTPResponseCode.BadRequest);
+  }
+}
+
+enum class ComputationType { TrainDescriptiveModel, TrainDiscriminantModel, ComputeAccuracy };
+
+// The four-dimensional Gaussian with all radii of one has the normalization coefficient of `pi^2`:
+//
+// * the definite integral of `exp(-(x*x + y*y + z*z + t*t)) / (pi^2)` is ``1.
```
Comment (mzhurovich) [inline examples/Iris/step3_optimize.cc]: ``1 -> \`1\`

File: examples/Iris/step3_optimize.cc [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+// * Ref.: http://www.wolframalpha.com/input/?i=exp(-(x*x%2By*y%2Bz*z%2Bt*t))%2F(pi^2)
+//
+// The four-dimensional Gaussian with the radii being `r` has the normalization coeffifient of `pi^2 * r^4)`.
```
Comment (mzhurovich) [inline examples/Iris/step3_optimize.cc]: Extra `)` after `r^4`?

Comment (mzhurovich) [inline examples/Iris/step3_optimize.cc]: s/coeffifient/coefficient/

File: examples/Iris/step3_optimize.cc [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+// * Ref.: http://www.wolframalpha.com/input/?i=exp(-((x/2)*(x/2)%2B(y/2)*(y/2)%2B(z/2)*(z/2)%2B(t/2)*(t/2)))/(pi^2*16)
+//
+// For all the radii being `r`, I move the coefficient under `exp(-(...)^2)`:
+// * Same integral is `1` for `exp(-(x*x+y*y+z*z+t*t) / (r^2)) / (pi^2*(r^4))`.
+// * Ref.: http://www.wolframalpha.com/input/?i=exp(-(x*x%2By*y%2Bz*z%2Bt*t)+%2F+(2^2))+%2F+(pi^2*(2^4))
+//         (Replace `2` by, say, `5`, i.e. { `2^2`, `2^4` -> `5^2`, `5^4` }, and the integral will stay `1`.
+//
+// Thus, as an implementation detail, model parameters to optimize, the per-coordinate radii, are stored
+// as `-log(sqr(r_i))`. This way, where `x` is the corresponding per-cluster radius-parameter:
+// * The negated exponent argument, which is `1/(r^2)`, is `inv_r_squared`, is `exp(x)`.
```
Comment (mzhurovich) [inline examples/Iris/step3_optimize.cc]: I don't quite get, what "which is `1/(r^2)`, is `inv_r_squared`, is `exp(x)" means exactly.

File: examples/Iris/step3_optimize.cc [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+
+    const T accuracy = fncas::exp(penalty_discriminant / valid_examples);
+
+    return fncas::optimize::ObjectiveFunctionValue<T>([&]() {
+      if (computation_type == ComputationType::ComputeAccuracy) {
+        // Return the probability, and the starting point should be `(1/3)`.
+        return accuracy;
+      } else {
+        // Just return the non-normalized number, for optimization purposes.
+        return computation_type == ComputationType::TrainDescriptiveModel ? penalty_descriptive : penalty_discriminant;
```
Comment (mzhurovich) [inline examples/Iris/step3_optimize.cc]: If you search for the maximum of the objective function, it's definitely not the penalty you want to maximize...

File: examples/Iris/step3_optimize.cc [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+          w[c] = computer.WeightInClass(flower, c, x);
+          ws += w[c];
+        }
+        penalty_descriptive += fncas::log(w[label_cit->second]);
+        penalty_discriminant += fncas::log(w[label_cit->second] / ws);
+      }
+    }
+    assert(valid_examples > 0u);
+
+    const T accuracy = fncas::exp(penalty_discriminant / valid_examples);
```
Comment (mzhurovich) [inline examples/Iris/step3_optimize.cc]: How does this accuracy value correspond to the intuitive concept of the single-label classification accuracy as the share of the correctly identified class label?

File: examples/Iris/step3_optimize.cc (hunk 4 above)
Comment (dkorolev) [inline examples/Iris/step3_optimize.cc]: :+1: 47b7744

File: examples/Iris/step3_optimize.cc (hunk 5 above)
Comment (dkorolev) [inline examples/Iris/step3_optimize.cc]: :+1: 47b7744

File: examples/Iris/step3_optimize.cc (hunk 6 above)
Comment (dkorolev) [inline examples/Iris/step3_optimize.cc]: :+1: :+1: 47b7744

File: examples/Iris/step3_optimize.cc (hunk 7 above)
Comment (dkorolev) [inline examples/Iris/step3_optimize.cc]: Rephrased, PTAL.

File: examples/Iris/step3_optimize.cc (hunk 8 above)
Comment (dkorolev) [inline examples/Iris/step3_optimize.cc]: It's negative, natural `log`-s of probabilities. Made it explicit in naming.

File: examples/Iris/step3_optimize.cc (hunk 9 above)
Comment (dkorolev) [inline examples/Iris/step3_optimize.cc]: Hmm, it's exactly that: The share of the correctly identified class label, its geometric mean over all the datapoints, computed as the exponent of the average value of the logarithms. Pls help me phrase it better!

**Comment by dkorolev** [comment]
@mzhurovich great eye, thank you! Fixed and rephrased everything, please take another look.

**Pull request merged**